### PR TITLE
RST-6016 - unittests to check elements on admin page

### PR DESCRIPTION
**Pull request by avi-singh-1912**
Flask unit tests to check elements on admin page

### CODE REVIEW

File: admin.Dockerfile [hunk 1]
```diff
@@ -34,4 +34,4 @@ RUN chmod 777 /app/run_admin.sh
 # Don't run as root user
 USER 1000
 CMD /app/run_admin.sh
-# CMD [ "python3", "-m" , "flask", "run", "--host=0.0.0.0" "--port=5001"]
\ No newline at end of file
+# CMD [ "python3", "-m" , "flask", "run", "--host=0.0.0.0" "--port=5001"]
```
Comment (itoulou) [inline admin.Dockerfile:37]: 🤔 ??

File: tests/admin/admin/test_page.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
+from tests.admin.helpers.data import create_test_applications
```
Comment (itoulou) [inline tests/admin/admin/test_page.py:1]: Believe this needs to be deleted,

File: admin.Dockerfile (hunk 1 above)
Comment (avi-singh-1912) [reply admin.Dockerfile:37]: sorry this was me accidentally committing admin.dockerfile and not knowing how to undo this file commit

File: tests/admin/conftest.py [hunk 3]
```diff
@@ -86,3 +88,46 @@ def expired_security_code(app):
 
         db.session.delete(security_code_)
         db.session.commit()
+
+
+@pytest.fixture()
+def submitted_application(app):
+    with app.app_context():
+        with app.test_request_context():
+            application = create_test_applications(status=ApplicationStatus.SUBMITTED, number_of_applications=1)
```
Comment (itoulou) [inline tests/admin/conftest.py:97]: Not quite. I wouldn't use this helper function. I would create multiple Application records similarly to how it's done in create_test_applications.

File: tests/admin/conftest.py [hunk 4]
```diff
@@ -86,3 +88,46 @@ def expired_security_code(app):
...
+            application = application[0]
```
Comment (itoulou) [inline tests/admin/conftest.py:98]: use .first() instead of indexing ([0])

File: tests/admin/applications/test_index.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    def test_homepage_elements(self, app, client):
+        with app.app_context():
+            with client.session_transaction() as session:
+                session['signedIn'] = 'test.email@example.com'
+
+            response = client.get('/applications')
+            html = response.data.decode()
+            assert response.status_code == 200
+            assert 'View and download GRC applications' in html
```
Comment (itoulou) [inline tests/admin/applications/test_index.py:11]: These assertions are a bit overkill

File: tests/admin/applications/test_index.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+            response = client.get('/applications#new')
+            html = response.data.decode()
+            assert response.status_code == 200
+            assert 'New applications' in html
+            assert 'View application' in html
+            assert 'Reference number' in html
+            assert 'Applicant name' in html
+            assert 'Submitted' in html
+            assert f'/applications/{submitted_application.reference_number}' in html
```
Comment (itoulou) [inline tests/admin/applications/test_index.py:30]: I think a good test would be to assert that the application data is present and correct in the headers. E.g. Headers in new applications is Reference number, applicant name and Submitted date and assert these are correct

File: admin.Dockerfile (hunk 1 above)
Comment (itoulou) [inline admin.Dockerfile:37]: 🤔 ??

File: admin.Dockerfile [hunk 7]
```diff
@@ -10,6 +10,7 @@ RUN wget https://github.com/jwilder/dockerize/releases/download/$DOCKERIZE_VERSI
 RUN curl -SLO https://deb.nodesource.com/nsolid_setup_deb.sh
 RUN chmod 500 nsolid_setup_deb.sh
 RUN ./nsolid_setup_deb.sh 21
+RUN apt-get update
```
Comment (itoulou) [inline admin.Dockerfile:13]: Please remove

File: tests/admin/applications/test_index.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            response = client.get('/applications')
+            html = response.data.decode()
+            assert 'Download GRC applications' in html
+
+    def test_new_applications_present(self, app, client, submitted_application):
+        with app.app_context():
+            with client.session_transaction() as session:
+                session['signedIn'] = 'test.email@example.com'
+
+            print('Firing up with:', submitted_application.application_data())
```
Comment (itoulou) [inline tests/admin/applications/test_index.py:17]: Do we need these logs?

File: tests/admin/applications/test_index.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            response = client.get('/applications#new')
+            html = response.data.decode()
+            assert 'New applications' in html
+            assert 'View application' in html
+            assert 'Reference number' in html
+            assert 'Applicant name' in html
+            assert 'Submitted' in html
+            assert '/applications/ABCD1234' in html
+            assert 'ABCD1234' in html
+            assert f'{submitted_application.user_input}' in html
```
Comment (itoulou) [inline tests/admin/applications/test_index.py:29]: Can you remove the user input from the template. It shouldn't be rendered in the template. Once thats done then  you don't need this assertion anymore

File: tests/admin/applications/test_index.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            assert f'{submitted_application.completed.strftime("%d/%m/%Y %H:%M")}' in html
```
Comment (itoulou) [inline tests/admin/applications/test_index.py:30]: Can you hard code this as well. Set the submission datetime in the fixture or in the test and then assert the value here.

File: tests/admin/conftest.py [hunk 11]
```diff
@@ -6,6 +6,9 @@
 from grc.models import db, AdminUser, SecurityCode
 from grc.utils.security_code import security_code_generator
 from werkzeug.security import generate_password_hash
+from tests.admin.helpers.data import create_test_applications
```
Comment (itoulou) [inline tests/admin/conftest.py:9]: Please remove if not being used.

File: tests/admin/conftest.py [hunk 12]
```diff
@@ -86,3 +88,149 @@ def expired_security_code(app):
...
+
+            # Commit the changes to the database
+            db.session.commit()
+
+            # Add user input
+            application_data = ApplicationData()
+            application_data.reference_number = reference_number
+            application_data.email_address = 'test.email@example.com'
+
+            user_input: str = jsonpickle.encode(application_data)
```
Comment (itoulou) [inline tests/admin/conftest.py:156]: For downloaded and completed app I think we can just use the save_application() method.

**Pull request closed without merging**